### ISSUE TITLE
SteamInput GetDigitalActionGlyph returns a null string.

### ISSUE THREAD

**Issue by ramunator** [Facepunch/Facepunch.Steamworks#708]
**Describe the bug**
GetDigitalActionGlyph returns a null string instead of the path to the icon.

**To Reproduce**
Steps to reproduce the behaviour: paste the code into your project maybe assign the keyboard icon if wanted

**Calling Code**
```
public static Texture2D LoadSteamGlyph(string digitalActionHandle, Texture2D keyboardIcon = null)
    {
        if (!SteamClient.IsLoggedOn)
        {
            Console.Log("Can't get icon while steam client is not running " + digitalActionHandle);
            return null;
        }

        SteamInput.RunFrame();

        if (SteamInput.Controllers.Count() != 0)
        {
            var controller = SteamInput.Controllers.ElementAt(0);
            
            Console.Log("Get icon for controller " + controller.Id + " path " + digitalActionHandle);
            
            string localGlyphPath = SteamInput.GetDigitalActionGlyph(controller, digitalActionHandle);

            if (string.IsNullOrEmpty(localGlyphPath))
            {
                Console.Log("Could not get icon of controller");
                return null;
            }

            Console.Log("Got icon of controller " + localGlyphPath);
            
            Texture2D iconTexture = LoadImage(localGlyphPath);
            return iconTexture;
        }
        else
        {
            Console.Log("No controller connected");
        }

        return keyboardIcon;
    }
```

**Expected behaviour**
I expect the function to return a Texture2D of the glyph relative to the handle

**Desktop (please complete the following information):**
 - OS: PC
- Unity: [e.g Unity 2019.3.18f]

**Additional context**
Hey, so this code should return the icon right? Nothing is returned. The controller id gets shown in the console. the localGlyphPath returns null for some reason. Please help I have no idea why this is happening. Also, the digitalActionHandle is correct and reading input works fine

**Comment by Peewi** [comment]
I'm surprised to hear that reading input is working. I found that it wasn't working because Steam Input never got initialized.

I only got it to work after making [a fork that calls the internal `Init` method](https://github.com/Facepunch/Facepunch.Steamworks/pull/702), but with that change `GetDigitalActionGlyph` works and returns a path to an image file.

**Comment by ramunator** [comment]
Yes indeed i Also had to decompile the dll add the Internal.init func but it does only make the controller detection work. What exactly have you changed and how do you get the icon

**Comment by Peewi** [comment]
In `SteamInput.InitializeInterface` I added a call to `Internal.Init( false )`. I have made no other changes to existing methods.

You say you decompiled the DLL. If you mean the DLL from the most recent release, it is 3 years old and there have been many code changes since then. It might make a difference if you build from the current source instead.

**Comment by ramunator** [comment]
Thanks for the answer. I used dotPeek to decompile some stuff had to be changed to get bo errors not sure if i removed some stuff Also i Think i set the init to false. Are you using dll files for your project or what are you doing?

**Comment by Peewi** [comment]
I am still using Facepunch.Steamworks in the form of a DLL. Just one I compiled, instead of an out-of-date release.

When doing this, remember to also use the latest Steamworks DLL (version 1.55).

**Comment by ramunator** [comment]
Thanks for all the help i dont Think i understand everything you said and it's not your problem just me being stupid is there any chance you could share your computer files?

**Comment by ramunator** [comment]
i figured some stuff out but i cant figure out how to build the DLL

**Comment by ramunator** [comment]
Im getting somewhere it all works in jetbrains rider but I get errors in unity saying namespace "Steamworks" does not exist. pretty strange. I will text back when I get a solution

**Comment by ramunator** [comment]
Fixed it.
For anyone seing a error like that you have to select the DLL and other files in the folder called net46 and not net6.0
![V0LQZKN](https://user-images.githubusercontent.com/91425663/223239363-0800e390-64b2-4b7f-8a4e-e89138f883df.png)

**Comment by ramunator** [comment]
After a while i figured it out.
1. Download the source code and not the release.
2. Go into the folder and open the file "Facepunch.Steamworks.Win64.csproj" in visual studio.
3. Right-click on the solution and hit build solution.
4. Drag the generated files from "Folder/Facepunch.Steamworks\bin\Debug\net46" 
make sure it's 4.6 and not 6.0, I just dragged the win64 files in because it gives errors when dragging more DLL's
5. Go into unity and create a script something like this one
6. `   
    public RawImage icon;

    private string cached = string.Empty;

    IEnumerator LoadTextureFromCache(string filePath)
    {
        if (!File.Exists(filePath))
        {
            yield break;
        }
        var www = UnityWebRequestTexture.GetTexture("file://" + filePath);
        yield return www.SendWebRequest();
        //texture loaded
        icon.texture = DownloadHandlerTexture.GetContent(www);
 
    }
    
    private void Update()
    {
        if (!string.IsNullOrEmpty(cached))
        {
            return;
        }
        Controller controller = SteamInput.Controllers.ElementAt(0);
        controller.ActionSet = "Action Set";
        cached = SteamInput.GetPngActionGlyph(controller, "Handle", GlyphSize.Small);

        if (string.IsNullOrEmpty(cached))
        {
            return;
            
        }
        
        Console.Log("Found controller icon, path: " + cached);
        StartCoroutine(LoadTextureFromCache(cached));
    }`

**Comment by stoora** [comment]
Hi I'm trying to detect joycon pair controllers. I edited and recompiled and referenced the dll correctly but controller collection always empty.
How did you manage to detect controllers correctly? 
```
SteamClient.Init(1456390);
SteamInput.Controllers.Count(); 👈 always 0
```

**Comment by ramunator** [comment]
well if they detect in steam and show up as connected in the steam client they should pop up after using init. You could try and log the steaminput.controllers count in update

**Comment by stoora** [comment]
Actually yes..., I'm double checking in the update method. 
I've just got it to work with a little hack.
I had to force 'openInterface' by bypassing the 'valid' check in "SteamClient.cs". and recompiled the assembly. 

`		internal static void AddInterface<T>() where T : SteamClass, new()
		{
			var t = new T();
			bool valid = t.InitializeInterface( false );
				openInterfaces.Add( t );
			//if ( valid )
			//{
			//	openInterfaces.Add( t );
			//}
			//else
			//{
			//	t.DestroyInterface( false );
			//}
		}`
		
this may not be the correct way to do it but it's my only solution for now.
